### PR TITLE
[16.0] [MIG] account_cash_invoice: Migration to 16.0

### PR DESCRIPTION
**Pull request by carlos-lopez-tecnativa**
- refactor to new reconciliation process(no more cash.box.out, cash.box.in
- no depends on account.bank.statement records, can use with account.bank.statement.line directly

@Tecnativa TT49793

### CODE REVIEW

File: account_cash_invoice/readme/CONTRIBUTORS.rst [hunk 1]
```diff
@@ -2,3 +2,6 @@
 * Jordi Ballester <jordi.ballester@eficent.com>
 * Jaime Arroyo <jaime.arroyo@creublanca.es>
 * Manuel Alejandro <buzondemam@gmail.com>
+* `Tecnativa <https://www.tecnativa.com>`_:
+
+    * Carlos Lopez<carlos.lopez@tecnativa.com>
```
Comment (pedrobaeza) [inline account_cash_invoice/readme/CONTRIBUTORS.rst]: Don't put the email for avoiding spam

```suggestion
    * Carlos López
```

File: account_cash_invoice/wizard/cash_pay_invoice.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                    else ["in_invoice", "out_refund"]
+                )
+                invoice_domain.append(("move_type", "in", move_type))
+            else:
+                # to avoid select invoice if invoice_type is not set
+                invoice_domain.append(("move_type", "in", []))
+            wizard.invoice_domain = invoice_domain
+
+    @api.onchange("invoice_type")
+    def _onchange_invoice_type(self):
```
Comment (pedrobaeza) [inline account_cash_invoice/wizard/cash_pay_invoice.py]: Replace onchanges by computed writable fields with the corresponding depends

File: account_cash_invoice/wizard/cash_pay_invoice.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    @api.onchange("invoice_id")
+    def _onchange_invoice(self):
+        if self.invoice_id:
+            self.amount = self.invoice_id.amount_residual_signed
+
+    def action_pay_invoice(self):
+        BankStatementLine = self.env["account.bank.statement.line"]
+        statement_line_vals = self._prepare_values_for_statement_line(self)
+        BankStatementLine.create(statement_line_vals)
+        return {"type": "ir.actions.act_window_close"}
```
Comment (pedrobaeza) [inline account_cash_invoice/wizard/cash_pay_invoice.py]: No need to explicitly passes such dictionary

File: account_cash_invoice/wizard/cash_pay_invoice.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def _prepare_values_for_statement_line(self, record):
```
Comment (pedrobaeza) [inline account_cash_invoice/wizard/cash_pay_invoice.py]: Convention for method names is:

```suggestion
    def _prepare_statement_line_vals(self):
```

And `record` is not needed. And if `record` would be another model record, the prepare method should be there in such model, not here, but it's not the case.

File: account_cash_invoice/wizard/cash_pay_invoice.xml [hunk 5]
```diff
@@ -1,42 +1,39 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <odoo>
-    <record id="cash_invoice_in_form" model="ir.ui.view">
-        <field name="name">cash_invoice_in_form</field>
-        <field name="model">cash.invoice.in</field>
+    <record id="cash_pay_invoice_form" model="ir.ui.view">
```
Comment (pedrobaeza) [inline account_cash_invoice/wizard/cash_pay_invoice.xml]: Convention for XML-IDs of views:

```suggestion
    <record id="view_cash_pay_invoice_form" model="ir.ui.view">
```

File: account_cash_invoice/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+This modules allows to pay an existing Supplier Invoice / Customer Refund, or
```
Comment (pedrobaeza) [inline account_cash_invoice/readme/DESCRIPTION.rst]:
```suggestion
This modules allows to pay existing customer invoices/refunds or vendor bills/refunds, or
```

File: account_cash_invoice/readme/ROADMAP.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+* Cannot pay invoices in a different currency than that defined in the journal
+  associated to the payment method used to pay/collect payment.
+* Cannot select a specific journal item on invoice with some lines(i.e. when use Payment terms)
```
Comment (pedrobaeza) [inline account_cash_invoice/readme/ROADMAP.rst]: Leave space before parenthesis:

```suggestion
* Cannot select a specific journal item on invoice with some lines (i.e. when use Payment terms)
```

File: account_cash_invoice/security/ir.model.access.csv [hunk 8]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_cash_pay_invoice,access_cash_pay_invoice,model_cash_pay_invoice,base.group_user,1,1,1,1
```
Comment (pedrobaeza) [inline account_cash_invoice/security/ir.model.access.csv]: Too broad permission. Please restrict to the real ones, which are those that do reconciliation.

**Review by carlos-lopez-tecnativa — COMMENTED** [review]
@pedrobaeza I have applied your suggestions. Could you please update your review?

File: account_cash_invoice/models/account_move.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (C) 2024 Tecnativa <https://tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    def _post(self, soft=True):
```
Comment (pedrobaeza) [inline account_cash_invoice/models/account_move.py]: Seeing this, please avoid this override that is anti-performant. Do the reconciliation directly from the wizard for the specific lines.

File: account_cash_invoice/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -1,2 +1,2 @@
-This modules allows to pay an existing Supplier Invoice / Customer Refund, or
+This modules allows to pay existing customer invoices/refunds or vendor bills/refunds, or
 to collect payment for an existing Customer Invoice from a Cash Statement.
```
Comment (pedrobaeza) [inline account_cash_invoice/readme/DESCRIPTION.rst]: Sorry, didn't read the rest of the description. I think this line can go out now.

File: account_cash_invoice/readme/DESCRIPTION.rst [hunk 11]
```diff
@@ -1,2 +1,2 @@
-This modules allows to pay an existing Supplier Invoice / Customer Refund, or
+This modules allows to pay existing customer invoices/refunds or vendor bills/refunds, or
```
Comment (pedrobaeza) [inline account_cash_invoice/readme/DESCRIPTION.rst]: s/This modules/This module

File: account_cash_invoice/wizard/cash_pay_invoice.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            if wizard.invoice_type:
+                move_type = ALLOWED_TYPES.get(wizard.invoice_type)
+                invoice_domain.append(("move_type", "in", move_type))
+            else:
+                # to avoid select invoice if invoice_type is not set
+                invoice_domain.append(("move_type", "in", []))
+            wizard.invoice_domain = invoice_domain
+
+    @api.depends("invoice_type")
+    def _compute_invoice(self):
```
Comment (pedrobaeza) [inline account_cash_invoice/wizard/cash_pay_invoice.py]: Proper convention:

```suggestion
    def _compute_invoice_id(self):
```

**Review by pedrobaeza — APPROVED** [review]

**Comment by carolinafernandez-tecnativa** [comment]
Please squash administrative commits.
Functional review LGTM.

**Review by carolinafernandez-tecnativa — APPROVED** [review]

**Pull request merged**